File: DSA-lab/modules/01-komplexitaetsanalyse-big-o/projects/03-final/run_audit.py
"""Benchmark the baseline and optimized customer aggregation pipelines."""

from __future__ import annotations

import argparse
import csv
from dataclasses import dataclass
from pathlib import Path
from statistics import median
from time import perf_counter
from typing import Callable, Sequence

from audit_pipeline import (
    CustomerSummary,
    Event,
    inefficient_pipeline,
    load_events,
    optimized_pipeline,
    write_summaries,
)
from data.generate_data import DEFAULT_SEED, generate_events

Pipeline = Callable[[Sequence[Event]], list[CustomerSummary]]

DEFAULT_ROWS = 8_000
DEFAULT_SIZES = (500, 1_000, 2_000, 4_000, 8_000)
PROJECT_DIR = Path(__file__).resolve().parent
DEFAULT_DATA_PATH = PROJECT_DIR / "data" / "events.csv"
DEFAULT_OUTPUT_DIR = PROJECT_DIR / "results"


@dataclass(frozen=True)
class AuditMeasurement:
    """Before/after timing for one input size."""

    size: int
    baseline_seconds: float
    optimized_seconds: float
    customer_count: int

    @property
    def speedup(self) -> float:
        """Return baseline runtime divided by optimized runtime."""
        return self.baseline_seconds / self.optimized_seconds


_AUDIT_SINK = 0


def median_runtime(
    pipeline: Pipeline,
    events: Sequence[Event],
    repeats: int,
) -> float:
    """Measure median runtime and retain the output length."""
    global _AUDIT_SINK

    durations = []
    for _ in range(repeats):
        start = perf_counter()
        summaries = pipeline(events)
        durations.append(perf_counter() - start)
        _AUDIT_SINK ^= len(summaries)
    return median(durations)


def validate_benchmark(
    events: Sequence[Event],
    sizes: Sequence[int],
    repeats: int,
) -> None:
    """Raise ValueError for an invalid audit configuration."""
    if not sizes:
        raise ValueError("at least one benchmark size is required")
    if any(size < 1 for size in sizes):
        raise ValueError("all benchmark sizes must be positive")
    if any(left >= right for left, right in zip(sizes, sizes[1:])):
        raise ValueError("benchmark sizes must be strictly increasing")
    if sizes[-1] > len(events):
        raise ValueError(
            f"largest benchmark size {sizes[-1]} exceeds "
            f"available rows {len(events)}"
        )
    if repeats < 1:
        raise ValueError("repeats must be positive")


def benchmark_pipelines(
    events: Sequence[Event],
    sizes: Sequence[int],
    repeats: int,
) -> list[AuditMeasurement]:
    """Verify equality and benchmark both implementations."""
    validate_benchmark(events, sizes, repeats)
    measurements = []

    for size in sizes:
        subset = events[:size]
        baseline_result = inefficient_pipeline(subset)
        optimized_result = optimized_pipeline(subset)
        if baseline_result != optimized_result:
            raise AssertionError(
                f"pipeline outputs differ for input size {size}"
            )

        baseline_seconds = median_runtime(
            inefficient_pipeline,
            subset,
            repeats,
        )
        optimized_seconds = median_runtime(
            optimized_pipeline,
            subset,
            repeats,
        )
        measurements.append(
            AuditMeasurement(
                size=size,
                baseline_seconds=baseline_seconds,
                optimized_seconds=optimized_seconds,
                customer_count=len(optimized_result),
            )
        )

    return measurements


def write_measurements(
    measurements: Sequence[AuditMeasurement],
    destination: Path,
) -> None:
    """Write before/after timings and speedups to CSV."""
    destination.parent.mkdir(parents=True, exist_ok=True)
    with destination.open("w", newline="", encoding="utf-8") as csv_file:
        writer = csv.writer(csv_file)
        writer.writerow(
            [
                "input_rows",
                "customer_count",
                "baseline_seconds",
                "optimized_seconds",
                "speedup",
            ]
        )
        for point in measurements:
            writer.writerow(
                [
                    point.size,
                    point.customer_count,
                    f"{point.baseline_seconds:.12g}",
                    f"{point.optimized_seconds:.12g}",
                    f"{point.speedup:.6f}",
                ]
            )


def write_report(
    measurements: Sequence[AuditMeasurement],
    destination: Path,
    seed: int,
    repeats: int,
) -> None:
    """Write a reproducible Markdown audit report."""
    destination.parent.mkdir(parents=True, exist_ok=True)
    largest = measurements[-1]
    lines = [
        "# Performance-Audit — Messergebnis",
        "",
        "## Konfiguration",
        "",
        f"- Seed: {seed}",
        f"- Wiederholungen pro Variante und Größe: {repeats}",
        "- Messgrenze: reine Aggregation; CSV-Erzeugung und Laden ausgeschlossen",
        "- Korrektheitsbedingung: vollständige Ergebnisgleichheit vor jeder Messung",
        "",
        "## Vorher/Nachher",
        "",
        "| Zeilen | Kunden | Baseline (s) | Optimiert (s) | Speedup |",
        "|---:|---:|---:|---:|---:|",
    ]
    for point in measurements:
        lines.append(
            f"| {point.size} | {point.customer_count} | "
            f"{point.baseline_seconds:.6f} | "
            f"{point.optimized_seconds:.6f} | {point.speedup:.2f}x |"
        )

    lines.extend(
        [
            "",
            "## Automatische Zusammenfassung",
            "",
            (
                f"Bei {largest.size} Ereignissen verarbeitet die optimierte "
                f"Pipeline dieselbe Ausgabe in {largest.optimized_seconds:.6f} "
                f"Sekunden statt {largest.baseline_seconds:.6f} Sekunden. "
                f"Das entspricht einem Speedup von {largest.speedup:.2f}x."
            ),
            "",
            "Die Baseline verwendet Listen-Membership und wiederholte "
            "Vollscans. Ihre Laufzeit ist O(nu) und damit bei u in O(n) "
            "quadratisch. Die optimierte Variante aggregiert per Hash Map in "
            "erwarteter O(n)-Zeit und sortiert u Ausgaben in O(u log u).",
            "",
            "## Eigene Interpretation",
            "",
            "Ergänze hier Verdopplungsfaktoren, praktische Relevanz, Grenzen "
            "des synthetischen Benchmarks und mögliche nächste Schritte.",
        ]
    )
    destination.write_text("\n".join(lines) + "\n", encoding="utf-8")


def create_plot(
    measurements: Sequence[AuditMeasurement],
    destination: Path,
) -> None:
    """Plot runtime growth and measured speedup."""
    import matplotlib

    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    destination.parent.mkdir(parents=True, exist_ok=True)
    sizes = [point.size for point in measurements]

    figure, (runtime_axis, speedup_axis) = plt.subplots(
        1,
        2,
        figsize=(13, 5.5),
    )
    runtime_axis.plot(
        sizes,
        [point.baseline_seconds for point in measurements],
        marker="o",
        linewidth=2,
        label="Baseline",
    )
    runtime_axis.plot(
        sizes,
        [point.optimized_seconds for point in measurements],
        marker="o",
        linewidth=2,
        label="Optimized",
    )
    runtime_axis.set_xscale("log", base=2)
    runtime_axis.set_yscale("log")
    runtime_axis.set_xlabel("Input rows n")
    runtime_axis.set_ylabel("Median runtime (seconds)")
    runtime_axis.set_title("Before/after runtime growth")
    runtime_axis.grid(True, which="both", linestyle=":", alpha=0.65)
    runtime_axis.legend()

    speedup_axis.plot(
        sizes,
        [point.speedup for point in measurements],
        marker="o",
        color="tab:green",
        linewidth=2,
    )
    speedup_axis.set_xscale("log", base=2)
    speedup_axis.set_xlabel("Input rows n")
    speedup_axis.set_ylabel("Speedup factor")
    speedup_axis.set_title("Measured speedup")
    speedup_axis.grid(True, which="both", linestyle=":", alpha=0.65)

    figure.suptitle("Customer aggregation performance audit")
    figure.tight_layout()
    figure.savefig(destination, dpi=160)
    plt.close(figure)


def print_summary(measurements: Sequence[AuditMeasurement]) -> None:
    """Print the audit table to the terminal."""
    print("\nPerformance audit")
    print("=" * 78)
    print(
        f"{'rows':>8} {'customers':>10} {'baseline':>14} "
        f"{'optimized':>14} {'speedup':>10}"
    )
    for point in measurements:
        print(
            f"{point.size:8d} {point.customer_count:10d} "
            f"{point.baseline_seconds:13.6f}s "
            f"{point.optimized_seconds:13.6f}s "
            f"{point.speedup:9.2f}x"
        )


def parse_arguments() -> argparse.Namespace:
    """Parse command-line options."""
    parser = argparse.ArgumentParser(
        description="Audit a deliberately inefficient aggregation pipeline."
    )
    parser.add_argument("--rows", type=int, default=DEFAULT_ROWS)
    parser.add_argument(
        "--sizes",
        type=int,
        nargs="+",
        default=list(DEFAULT_SIZES),
    )
    parser.add_argument("--repeats", type=int, default=3)
    parser.add_argument("--seed", type=int, default=DEFAULT_SEED)
    parser.add_argument("--data-path", type=Path, default=DEFAULT_DATA_PATH)
    parser.add_argument("--output-dir", type=Path, default=DEFAULT_OUTPUT_DIR)
    parser.add_argument(
        "--reuse-data",
        action="store_true",
        help="reuse data-path instead of regenerating it",
    )
    return parser.parse_args()


def main() -> None:
    """Generate data, run the audit, and persist all artifacts."""
    arguments = parse_arguments()
    if arguments.rows < 1:
        raise SystemExit("Configuration error: rows must be positive")

    if not arguments.reuse_data:
        generate_events(arguments.data_path, arguments.rows, arguments.seed)
        print(
            f"Generated {arguments.rows} events with seed {arguments.seed} "
            f"at {arguments.data_path}."
        )
    elif not arguments.data_path.exists():
        raise SystemExit(
            f"Configuration error: data file does not exist: "
            f"{arguments.data_path}"
        )

    events = load_events(arguments.data_path)
    try:
        measurements = benchmark_pipelines(
            events,
            arguments.sizes,
            arguments.repeats,
        )
    except ValueError as error:
        raise SystemExit(f"Configuration error: {error}") from error

    output_dir = arguments.output_dir
    summary_path = output_dir / "customer_summary.csv"
    measurements_path = output_dir / "performance_audit.csv"
    plot_path = output_dir / "performance_audit.png"
    report_path = output_dir / "AUDIT_REPORT.md"

    summaries = optimized_pipeline(events)
    write_summaries(summaries, summary_path)
    write_measurements(measurements, measurements_path)
    create_plot(measurements, plot_path)
    write_report(
        measurements,
        report_path,
        arguments.seed,
        arguments.repeats,
    )
    print_summary(measurements)

    print("\nArtifacts")
    for artifact in (
        summary_path,
        measurements_path,
        plot_path,
        report_path,
    ):
        print(f"  {artifact}")


if __name__ == "__main__":
    main()

File: DSA-lab-java/modules/01-komplexitaetsanalyse-big-o/projects/03-final/src/main/java/lab/dsa/module01/finalproject/RunAudit.java
package lab.dsa.module01.finalproject;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.image.BufferedImage;
import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import javax.imageio.ImageIO;
import lab.dsa.module01.finalproject.AuditPipeline.CustomerSummary;
import lab.dsa.module01.finalproject.AuditPipeline.Event;

/** Benchmarks the baseline and optimized customer aggregation pipelines. */
public final class RunAudit {
    public static final int DEFAULT_ROWS = 8_000;
    public static final List<Integer> DEFAULT_SIZES = List.of(500, 1_000, 2_000, 4_000, 8_000);
    private static volatile int auditSink;

    private RunAudit() {
    }

    @FunctionalInterface
    public interface Pipeline {
        List<CustomerSummary> apply(List<Event> events);
    }

    public record AuditMeasurement(int size, double baselineSeconds,
                                   double optimizedSeconds, int customerCount) {
        public double speedup() {
            return baselineSeconds / optimizedSeconds;
        }
    }

    private record Arguments(int rows, List<Integer> sizes, int repeats, long seed,
                             Path dataPath, Path outputDirectory, boolean reuseData, boolean help) {
    }

    public static double medianRuntime(Pipeline pipeline, List<Event> events, int repeats) {
        double[] durations = new double[repeats];
        for (int repeat = 0; repeat < repeats; repeat++) {
            long start = System.nanoTime();
            List<CustomerSummary> summaries = pipeline.apply(events);
            durations[repeat] = (System.nanoTime() - start) / 1_000_000_000.0;
            auditSink ^= summaries.size();
        }
        Arrays.sort(durations);
        return durations.length % 2 == 1
                ? durations[durations.length / 2]
                : (durations[durations.length / 2 - 1] + durations[durations.length / 2]) / 2;
    }

    public static void validateBenchmark(List<Event> events, List<Integer> sizes, int repeats) {
        if (sizes.isEmpty()) {
            throw new IllegalArgumentException("At least one benchmark size is required");
        }
        if (sizes.stream().anyMatch(size -> size < 1)) {
            throw new IllegalArgumentException("All benchmark sizes must be positive");
        }
        for (int index = 1; index < sizes.size(); index++) {
            if (sizes.get(index - 1) >= sizes.get(index)) {
                throw new IllegalArgumentException("Benchmark sizes must be strictly increasing");
            }
        }
        if (sizes.getLast() > events.size()) {
            throw new IllegalArgumentException("Largest benchmark size exceeds available rows");
        }
        if (repeats < 1) {
            throw new IllegalArgumentException("Repeats must be positive");
        }
    }

    public static List<AuditMeasurement> benchmarkPipelines(
            List<Event> events, List<Integer> sizes, int repeats) {
        validateBenchmark(events, sizes, repeats);
        List<AuditMeasurement> measurements = new ArrayList<>();
        for (int size : sizes) {
            List<Event> subset = events.subList(0, size);
            List<CustomerSummary> baselineResult = AuditPipeline.inefficientPipeline(subset);
            List<CustomerSummary> optimizedResult = AuditPipeline.optimizedPipeline(subset);
            if (!baselineResult.equals(optimizedResult)) {
                throw new AssertionError("Pipeline outputs differ for input size " + size);
            }
            double baselineSeconds = medianRuntime(AuditPipeline::inefficientPipeline, subset, repeats);
            double optimizedSeconds = medianRuntime(AuditPipeline::optimizedPipeline, subset, repeats);
            measurements.add(new AuditMeasurement(
                    size, baselineSeconds, optimizedSeconds, optimizedResult.size()));
        }
        return measurements;
    }

    public static void writeMeasurements(List<AuditMeasurement> measurements, Path destination)
            throws IOException {
        if (destination.getParent() != null) {
            Files.createDirectories(destination.getParent());
        }
        try (BufferedWriter writer = Files.newBufferedWriter(destination, StandardCharsets.UTF_8)) {
            writer.write("input_rows,customer_count,baseline_seconds,optimized_seconds,speedup");
            writer.newLine();
            for (AuditMeasurement point : measurements) {
                writer.write(String.format(Locale.ROOT, "%d,%d,%.12g,%.12g,%.6f",
                        point.size(), point.customerCount(), point.baselineSeconds(),
                        point.optimizedSeconds(), point.speedup()));
                writer.newLine();
            }
        }
    }

    public static void writeReport(List<AuditMeasurement> measurements, Path destination,
                                   long seed, int repeats) throws IOException {
        if (destination.getParent() != null) {
            Files.createDirectories(destination.getParent());
        }
        AuditMeasurement largest = measurements.getLast();
        List<String> lines = new ArrayList<>(List.of(
                "# Performance Audit — Measurement Results", "", "## Configuration", "",
                "- Seed: " + seed,
                "- Repetitions per variant and size: " + repeats,
                "- Measurement boundary: aggregation only; CSV generation and loading excluded",
                "- Correctness condition: complete output equality before every measurement",
                "", "## Before and After", "",
                "| Rows | Customers | Baseline (s) | Optimized (s) | Speedup |",
                "|---:|---:|---:|---:|---:|"));
        for (AuditMeasurement point : measurements) {
            lines.add(String.format(Locale.ROOT, "| %d | %d | %.6f | %.6f | %.2fx |",
                    point.size(), point.customerCount(), point.baselineSeconds(),
                    point.optimizedSeconds(), point.speedup()));
        }
        lines.addAll(List.of("", "## Automatic Summary", "",
                String.format(Locale.ROOT,
                        "At %,d events, the optimized pipeline produces the same output in %.6f seconds instead of %.6f seconds, a %.2fx speedup.",
                        largest.size(), largest.optimizedSeconds(), largest.baselineSeconds(), largest.speedup()),
                "", "The baseline uses list membership and repeated full scans, giving O(nu) time and O(n²) when u is O(n). The optimized version aggregates with a hash map in expected O(n) time and sorts u outputs in O(u log u).",
                "", "## Your Interpretation", "",
                "Add doubling factors, practical relevance, limitations of the synthetic benchmark, and possible next steps."));
        Files.write(destination, lines, StandardCharsets.UTF_8);
    }

    /** Creates a dependency-free PNG with runtime and speedup panels. */
    public static void createPlot(List<AuditMeasurement> measurements, Path destination) throws IOException {
        int width = 1_300;
        int height = 560;
        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D graphics = image.createGraphics();
        graphics.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        graphics.setColor(Color.WHITE);
        graphics.fillRect(0, 0, width, height);
        graphics.setColor(Color.BLACK);
        graphics.drawString("Customer aggregation performance audit", 520, 25);
        drawRuntimePanel(graphics, measurements, 55, 55, 570, 440);
        drawSpeedupPanel(graphics, measurements, 680, 55, 570, 440);
        graphics.dispose();
        if (destination.getParent() != null) {
            Files.createDirectories(destination.getParent());
        }
        ImageIO.write(image, "png", destination.toFile());
    }

    private static void drawRuntimePanel(Graphics2D graphics, List<AuditMeasurement> points,
                                         int x, int y, int width, int height) {
        graphics.setColor(Color.BLACK);
        graphics.drawString("Before/after runtime growth", x + 190, y);
        graphics.drawRect(x, y + 20, width, height - 20);
        double minimum = points.stream().mapToDouble(AuditMeasurement::optimizedSeconds).min().orElseThrow();
        double maximum = points.stream().mapToDouble(AuditMeasurement::baselineSeconds).max().orElseThrow();
        plotLine(graphics, points, x, y + 20, width, height - 20, minimum, maximum, true, Color.RED);
        plotLine(graphics, points, x, y + 20, width, height - 20, minimum, maximum, false, new Color(0, 140, 0));
        graphics.setColor(Color.RED); graphics.drawString("Baseline", x + 15, y + 45);
        graphics.setColor(new Color(0, 140, 0)); graphics.drawString("Optimized", x + 15, y + 65);
    }

    private static void plotLine(Graphics2D graphics, List<AuditMeasurement> points,
                                 int x, int y, int width, int height, double minimum,
                                 double maximum, boolean baseline, Color color) {
        graphics.setColor(color);
        graphics.setStroke(new BasicStroke(2));
        int previousX = -1;
        int previousY = -1;
        double logMinimum = Math.log(Math.max(minimum, 1e-12));
        double logMaximum = Math.log(Math.max(maximum, minimum * 1.01));
        for (int index = 0; index < points.size(); index++) {
            AuditMeasurement point = points.get(index);
            double value = baseline ? point.baselineSeconds() : point.optimizedSeconds();
            int currentX = x + index * width / Math.max(1, points.size() - 1);
            int currentY = y + height - (int) ((Math.log(value) - logMinimum)
                    / (logMaximum - logMinimum) * height);
            if (previousX >= 0) graphics.drawLine(previousX, previousY, currentX, currentY);
            graphics.fillOval(currentX - 4, currentY - 4, 8, 8);
            previousX = currentX; previousY = currentY;
        }
    }

    private static void drawSpeedupPanel(Graphics2D graphics, List<AuditMeasurement> points,
                                         int x, int y, int width, int height) {
        graphics.setColor(Color.BLACK);
        graphics.drawString("Measured speedup", x + 220, y);
        graphics.drawRect(x, y + 20, width, height - 20);
        double maximum = Math.max(1, points.stream().mapToDouble(AuditMeasurement::speedup).max().orElseThrow());
        graphics.setColor(new Color(0, 120, 0));
        graphics.setStroke(new BasicStroke(2));
        int previousX = -1;
        int previousY = -1;
        for (int index = 0; index < points.size(); index++) {
            int currentX = x + index * width / Math.max(1, points.size() - 1);
            int currentY = y + 20 + height - 20 - (int) (points.get(index).speedup() / maximum * (height - 20));
            if (previousX >= 0) graphics.drawLine(previousX, previousY, currentX, currentY);
            graphics.fillOval(currentX - 4, currentY - 4, 8, 8);
            previousX = currentX; previousY = currentY;
        }
    }

    public static void printSummary(List<AuditMeasurement> measurements) {
        System.out.println("\nPerformance audit");
        System.out.println("=".repeat(78));
        System.out.printf("%8s %10s %14s %14s %10s%n", "rows", "customers", "baseline", "optimized", "speedup");
        for (AuditMeasurement point : measurements) {
            System.out.printf(Locale.ROOT, "%8d %10d %13.6fs %13.6fs %9.2fx%n",
                    point.size(), point.customerCount(), point.baselineSeconds(), point.optimizedSeconds(), point.speedup());
        }
    }

    private static Arguments parseArguments(String[] arguments) {
        int rows = DEFAULT_ROWS;
        List<Integer> sizes = new ArrayList<>(DEFAULT_SIZES);
        int repeats = 3;
        long seed = GenerateData.DEFAULT_SEED;
        Path dataPath = Paths.get("data", "events.csv");
        Path outputDirectory = Paths.get("results");
        boolean reuseData = false;
        boolean help = false;
        for (int index = 0; index < arguments.length; index++) {
            switch (arguments[index]) {
                case "--rows" -> rows = Integer.parseInt(arguments[++index]);
                case "--repeats" -> repeats = Integer.parseInt(arguments[++index]);
                case "--seed" -> seed = Long.parseLong(arguments[++index]);
                case "--data-path" -> dataPath = Paths.get(arguments[++index]);
                case "--output-dir" -> outputDirectory = Paths.get(arguments[++index]);
                case "--reuse-data" -> reuseData = true;
                case "--help" -> help = true;
                case "--sizes" -> {
                    sizes = new ArrayList<>();
                    while (index + 1 < arguments.length && !arguments[index + 1].startsWith("--")) {
                        sizes.add(Integer.parseInt(arguments[++index]));
                    }
                }
                default -> throw new IllegalArgumentException("Unknown option: " + arguments[index]);
            }
        }
        return new Arguments(rows, sizes, repeats, seed, dataPath, outputDirectory, reuseData, help);
    }

    public static void main(String[] commandLineArguments) throws IOException {
        Arguments arguments = parseArguments(commandLineArguments);
        if (arguments.help()) {
            System.out.println("Options: --rows N --sizes N... --repeats N --seed N --data-path PATH --output-dir PATH --reuse-data");
            return;
        }
        if (arguments.rows() < 1) throw new IllegalArgumentException("Rows must be positive");
        if (!arguments.reuseData()) {
            GenerateData.generateEvents(arguments.dataPath(), arguments.rows(), arguments.seed());
            System.out.printf("Generated %d events with seed %d at %s.%n",
                    arguments.rows(), arguments.seed(), arguments.dataPath());
        } else if (!Files.exists(arguments.dataPath())) {
            throw new IllegalArgumentException("Data file does not exist: " + arguments.dataPath());
        }
        List<Event> events = AuditPipeline.loadEvents(arguments.dataPath());
        List<AuditMeasurement> measurements = benchmarkPipelines(events, arguments.sizes(), arguments.repeats());
        Path summaryPath = arguments.outputDirectory().resolve("customer_summary.csv");
        Path measurementsPath = arguments.outputDirectory().resolve("performance_audit.csv");
        Path plotPath = arguments.outputDirectory().resolve("performance_audit.png");
        Path reportPath = arguments.outputDirectory().resolve("AUDIT_REPORT.md");
        AuditPipeline.writeSummaries(AuditPipeline.optimizedPipeline(events), summaryPath);
        writeMeasurements(measurements, measurementsPath);
        createPlot(measurements, plotPath);
        writeReport(measurements, reportPath, arguments.seed(), arguments.repeats());
        printSummary(measurements);
        System.out.println("\nArtifacts\n  " + summaryPath + "\n  " + measurementsPath + "\n  " + plotPath + "\n  " + reportPath);
    }
}
